### PR TITLE
fix PreferenceChangeEvent<T> type

### PR DESCRIPTION
**Pull request by marechal-p**
#### What it does

Improves the `PreferenceChangeEvent<T>` type to properly reflect how it is meant to be used. I also changed the few places where the type was given wrong information. See commit message for details.

#### How to test

You can inspect the type of `event.newValue` in places where we check `event.preferenceName` and make sure that instead of inferring an union of all values for all preference names, we now have the proper type.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by marechal-p** [comment]
I noticed this oddity while reviewing something else, figured we could fix it.

File: packages/core/src/browser/preferences/preference-proxy.ts [hunk 1]
```diff
@@ -21,12 +21,61 @@ import { PreferenceService } from './preference-service';
...
     affects(resourceUri?: string, overrideIdentifier?: string): boolean;
-}
+} & {
+    [K in keyof T]-?: {
+        readonly preferenceName: K;
+        readonly newValue: T[K];
+        /**
+         * Undefined if the preference is set for the first time.
+         */
+        // TODO: Use the default value instead of undefined?
```
Comment (vince-fugnitto) [inline packages/core/src/browser/preferences/preference-proxy.ts:75]: Is there something we need to address for `oldValue` as part of this pull-request?

File: packages/workspace/src/browser/workspace-service.ts [hunk 2]
```diff
@@ -95,8 +95,8 @@ export class WorkspaceService implements FrontendApplicationContribution {
                 this.updateWorkspace();
             }
         });
-        this.fsPreferences.onPreferenceChanged(e => {
-            if (e.preferenceName === 'files.watcherExclude') {
+        this.fsPreferences.onPreferenceChanged(event => {
+            if (event.preferenceName === 'files.watcherExclude') {
```
Comment (vince-fugnitto) [inline packages/workspace/src/browser/workspace-service.ts:99]: I believe the change is unnecessary, in other places (earlier in the pr we have `e`). If we were to make the update we should probably follow one format.

File: packages/git/src/browser/diff/git-diff-widget.tsx [hunk 3]
```diff
@@ -76,9 +76,9 @@ export class GitDiffWidget extends BaseWidget implements StatefulWidget {
         this.containerLayout.addWidget(this.resourceWidget);
 
         this.updateViewMode(this.scmPreferences.get('scm.defaultViewMode'));
-        this.toDispose.push(this.scmPreferences.onPreferenceChanged((e: PreferenceChangeEvent<ScmConfiguration>) => {
+        this.toDispose.push(this.scmPreferences.onPreferenceChanged(e => {
```
Comment (vince-fugnitto) [inline packages/git/src/browser/diff/git-diff-widget.tsx:79]: Is there a reason why we no longer want to be explicit with the typing?

File: packages/workspace/src/browser/workspace-service.ts (hunk 2 above)
Comment (marechal-p) [reply packages/workspace/src/browser/workspace-service.ts:99]: Just aligned with another listener from the same file, I did not touch other places for the reason you mentioned.

File: packages/git/src/browser/diff/git-diff-widget.tsx (hunk 3 above)
Comment (marechal-p) [reply packages/git/src/browser/diff/git-diff-widget.tsx:79]: It's just noise in the code base in those cases. Although things like `function (event: ...): ...` still require it.

File: packages/core/src/browser/preferences/preference-proxy.ts (hunk 1 above)
Comment (marechal-p) [reply packages/core/src/browser/preferences/preference-proxy.ts:75]: I don't plan to, only `newValue` seemed oddly typed, `oldValue` looks more complex to understand what we would get as a value: the default or `undefined`?

**Review by vince-fugnitto — APPROVED** [review]
The changes look good to me, and fix the issue I've noticed when dealing with the `onPreferenceChanged` event among other things.

We now have the correct typing when doing the following and use the event directly:

```ts
this.preferences.onPreferenceChanged(event =>  {
    if (event.preferenceName === 'git.editor.dirtyDiff.linesLimit') {
        const value = event.newValue; // we correctly have a number.
    }
});
```

rather than the workaround for _master_:

```ts
this.preferences.onPreferenceChanged(event =>  {
    if (event.preferenceName === 'git.editor.dirtyDiff.linesLimit') {
        const newValue = event.newValue; // newValue does not return the proper typings (returns all typings from the schema).
        const value = this.preferences['git.editor.dirtyDiff.linesLimit']; // workaround to get the proper newValue.
    }
});
```

**Pull request merged**